File: protocol/src/main/resources/php/ProtocolRegistrationTemplate.php
class ${protocol_name}Registration implements IProtocolRegistration {
    public function protocolId(): int
    {
        return ${protocol_id};
    }

    public function write(ByteBuffer $buffer, mixed $packet): void
    {
        if ($packet == null)
        {
            $buffer->writeInt(0);
            return;
        }
        ${protocol_write_serialization}
    }

    public function read(ByteBuffer $buffer): mixed
    {
        $length = $buffer->readInt();
        $packet = new ${protocol_name}();
        if ($length == 0) {
            return $packet;
        }
        $beforeReadIndex = $buffer->readOffset();
        ${protocol_read_deserialization}
        if ($length > 0) {
            $buffer->setReadOffset($beforeReadIndex + $length);
        }
        return $packet;
    }
}

File: protocol/src/main/resources/dart/ProtocolRegistrationTemplate.dart
class ${protocol_name}Registration implements IProtocolRegistration<${protocol_name}> {
  @override
  int protocolId() {
    return ${protocol_id};
  }

  @override
  void write(IByteBuffer buffer, ${protocol_name}? packet) {
    if (packet == null) {
      buffer.writeInt(0);
      return;
    }
    ${protocol_write_serialization}
  }


  @override
  ${protocol_name} read(IByteBuffer buffer) {
    var length = buffer.readInt();
    var packet = ${protocol_name}();
    if (length == 0) {
      return packet;
    }
    var beforeReadIndex = buffer.getReadOffset();
    ${protocol_read_deserialization}
    if (length > 0) {
      buffer.setReadOffset(beforeReadIndex + length);
    }
    return packet;
  }
}